### PR TITLE
OCSP responder to serve status check for itself using latest CRL

### PR DESCRIPTION
**Pull request by fmarco76**
Modify OCSP responder to perform peer certificate checks with the CRL instead of using OCSP, avoiding the chicken and egg issue.

Solve RHCS-4262

### CODE REVIEW

**Comment by fmarco76** [comment]
The first step of inverting the init order to load the store first is done.

File: base/server/src/main/java/com/netscape/cmscore/apps/CMS.java
Comment (edewata) [inline base/server/src/main/java/com/netscape/cmscore/apps/CMS.java:57]: In the newer branches we've removed the dependency on global variables in `CMS` class. I'd suggest moving this variable into `CMSEngine` class.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java
Comment (edewata) [inline base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java:46]: The `.toString()` can be removed since it's implicitly called by string concatenation.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java
Comment (edewata) [inline base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java:58]: To simplify troubleshooting I'd suggest displaying the exception message and optionally attach the exception (to show the stack trace):
```
logger.error("...<message>...: " + e.getMessage(), e);
```

File: base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java
Comment (edewata) [inline base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java:148]: IIUC this loop will return `false` as soon as it finds the first valid cert. Should we check all certs in the chain instead?

**Review by edewata — COMMENTED** [review]
Please see my comments below. I have one question about the loop, and the rest are just minor suggestions.

**Review by edewata — APPROVED** [review]
Thanks for the update! Looks good to me, but I'll defer to others for the final approval.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java
Comment (ladycfu) [inline base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java:53]: I'm not certain about issuerDN string comparison here.  If you look at the processRequest(...) method in ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java, it uses the CA's public key hash as key to locate the right CRL, which is more assuring, isn't it?

File: base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java
Comment (ladycfu) [inline base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java:162]: same comment as above regarding using CA's public key hash.

**Review by ladycfu — COMMENTED** [review]
I have conducted a few tests with the PR on revoked DS server-cert, agent cert at webUI, and admin cert with pkiconsole and I think they work as expected.  Of course more testing could be used for other scenarios but I'll leave that to QE.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java
Comment (edewata) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java:142]: This code is only used by the OCSP subsystem. Is there any case where we want this functionality disabled? If it's always going to be enabled maybe we don't need this param.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java
Comment (fmarco76) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java:53]: I have implemented this using the [CheckCertServlet](https://github.com/dogtagpki/pki/blob/v10.13/base/ocsp/src/main/java/com/netscape/cms/servlet/ocsp/CheckCertServlet.java#L167) as reference where the `ICRLIssuingPointRecord` is identified using the common name. Nevertheless, using the key hash is definitely a stronger mechanism. I will try to use the `processRequest` method or extract some code in a separate method.

File: base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java
Comment (fmarco76) [reply base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java:162]: Same approach as above.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java
Comment (fmarco76) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java:142]: I agree with @edewata, I do not see a reason why this should be disabled so I think we could make permanent this change and not triggered by an option. However, we could just makes it the default behaviour just in case the scenario where disabling the check arise. @ladycfu I am converting to `true`.

I have some concerns with `auths.revocationChecking.enabled=true`. For the OCSP subsystem this work only if the CA sends CRLs through an LDAP and not with direct publishing. Should we stopped, or at least log an error, if it is true but LDAP publishing is not configured?

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java
Comment (fmarco76) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java:59]: @ladycfu I did not find an easy way to get the issuer  key hash from the certificate but to be more confident I have verified that the CA cert is the one signing the certificate. The test is done only if the DNs match to avoid the overhead of encoding/decoding certificates. I have performed the operation considering _Mozilla-JSS_ provider is available. If this is not always the case some extra step are needed. Is this approach viable for the job?

**Comment by fmarco76** [comment]
With last commit there is an error on TPS  tests but  I do not see changes in this PR that could generate the error

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java
Comment (ladycfu) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java]: The reason why I propose having a "switch" to go back to the previous behavior was because we want to be sure that there is no gotcha's in the new mechanism.  For now, I think this switch serves its purpose, as a back up .  I think it's fine to have it enabled by default.   Now, wrt auths.revocationChecking.enabled=true, this "Method C" precludes the direct CA->OCSP CRL publishing because the ocsp's ssl port is not able to receive CRL from the CA due to chick-n-egg issue.  The CA depends on the ocsp to start up, and the ocsp depends on the CA to start up, if replying on ca->ocsp direct publishing. The CA is to have enableOCSP=true and its internaldb's  server-cert is to bear an AIA pointing to the ocsp.  @fmarco76 I think you have a good suggestion there on checking whether it's ldapStore that's being set for crl update if ocsp.store.ldapStore.checkSubsystemConnection=true.  btw, what is the concern you wrt auths.revocationChecking.enabled=true ?

Comment (ladycfu) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java]: We need to conduct tests to make sure both ocsp and cap can start in the order of ocsp and then ca.  In the setup, the CA should have its enableOCSP=true in server.xml with both ocspResponderURL and ocspResponderCertNickname removed.  CA's internaldb should have its server-cert bear AIA pointing to the ocsp (instead of the default internal ocsp of the ca).

Comment (ladycfu) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java]: @fmarco76 btw, why do you call this flag  "checkSubsystemConnection"?  To me it seems something like "ocsp.store.ldapStore.validateConnCertWithCRL" would be more telling.  Actually, could you add an explanation in comment where this flag is defined what exactly it does?  We need to add something in the doc.  Thanks.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java
Comment (ladycfu) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java:59]: Hi @fmarco76 , I think in general, it's a better practice to avoid string comparison when it comes to DN names, due to encoding differences.   Looking closely, the processRequest was for processing an ocsp request which contains the CertID, so it's not exactly what we have here.  We have the peer cert, X509Certificate certificate, and we have the ca cert, X509CertImpl ca_certImpl = new X509CertImpl(tPt.getCACert());  and we can get the ca's public key PublicKey ca_pubkey = ca_certImpl.getPublicKey(); where PublicKey is java.security.PublicKey.  Meanwhile, the peer cert X509Certificate certificate can have it's public key gotten with PublicKey peerCert_pubkey = certificate.getPublicKey(); So now you can compare. I'm not sure where you said you did signature verification.  I could be wrong, but that sounds more expensive than the PublicKey comparison.  Anyway, I think this patch is as good as it gets for the time we have, so you could leave this as an improvement for later if you want if you donn't have time for it now.  Thanks.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java
Comment (fmarco76) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java]: > I think you have a good suggestion there on checking whether it's ldapStore that's being set for crl update if ocsp.store.ldapStore.checkSubsystemConnection=true. 

This setting is done on ldapStore of the OCSP so we do not need to verify if it is used because if it is not then this setting does not apply.

> what is the concern you wrt auths.revocationChecking.enabled=true ?

In the OCSP, if this is true but the ldapStore is not used the revocation cannot be verified. I have not tested if this generate errors. Therefore I was wondering if the user should be notified that revocation will not work even it is enabled.

> why do you call this flag "checkSubsystemConnection"? To me it seems something like "ocsp.store.ldapStore.validateConnCertWithCRL" would be more telling. Actually, could you add an explanation in comment where this flag is defined what exactly it does? We need to add something in the doc. 

Yes, sound better, I will change the name. Where exactly should I write an explanation? A comment in github where the code is provided  or a comment inside the code?

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java
Comment (fmarco76) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/CRLLdapValidator.java:59]: > Hi @fmarco76 , I think in general, it's a better practice to avoid string comparison when it comes to DN names, due to encoding differences. 

IIRC String in java have the same encoding independently of how they are stored in the certificate. The encoding is converted when they are read/write. The format could be different but this is related to the used classes so it is known (although it could change moving to different versions of Java).

> Looking closely, the processRequest was for processing an ocsp request which contains the CertID, so it's not exactly what we have here. We have the peer cert, X509Certificate certificate, and we have the ca cert, X509CertImpl ca_certImpl = new X509CertImpl(tPt.getCACert()); and we can get the ca's public key PublicKey ca_pubkey = ca_certImpl.getPublicKey(); where PublicKey is java.security.PublicKey. Meanwhile, the peer cert X509Certificate certificate can have it's public key gotten with PublicKey peerCert_pubkey = certificate.getPublicKey(); So now you can compare. I'm not sure where you said you did signature verification. I could be wrong, but that sounds more expensive than the PublicKey comparison. 

The problem here is that you get the PublicKey of the CA and the PublicKey of the peer certificate. They are different so there is no way to compare. The OCSP request contain the hash of the issuer PublicKey so it is possible to get the hash of the CA PublicKey and compare. 
From the peer certificate I found no way to get the issuer PublicKey to compare with the CA.
The approach I implemented is in line [61-63 of CRLLdapValidator](https://github.com/dogtagpki/pki/pull/4504/files/d54adc9d9f11ac1dd31632c9cbd92fe65ab89a3a#diff-7b84e1e2f2129ea94596917fca937c8a365e0d697e3773a43e13d0f01ab122dd).

It is definitely more expensive of the PublicKey comparison and this is the reason I do only if the DN match. However, considering a low number of CA for an OCSP we could skip the DN match and do always the signature check. Is this assumption correct? 


> Anyway, I think this patch is as good as it gets for the time we have, so you could leave this as an improvement for later if you want if you don't have time for it now. Thanks.

I can work on this until the end of the sprint so we still have several days before merge.

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java
Comment (fmarco76) [inline base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java:84]: The option `ocsp.store.ldapStore.validateConnCertWithCRL` enables the revocation verification of peer certificates using the CRL stored in the LDAP shared with the CA.

When it is set to `true` (default value), the peer certificate of all the outcome connections from the OCSP subsystem are verified with the CRL.

If the option `auths.revocationChecking.enabled` is also set to `true` the peer certificate ot all the income connections to the OCSP subsystem are verified with the CRL.

Comment (ladycfu) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java:84]: How about put the info right here in the code where PROP_VALIDATE_CONNECTION_WITH_CRL = "validateConnCertWithCRL" is assigned?  Comments in code are longer lasting and provide info others later.  I will also be using it for the doc changes.

File: base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java
Comment (ladycfu) [inline base/ocsp/src/main/java/org/dogtagpki/server/ocsp/OCSPEngine.java:167]: May I suggest that you rename the variable "ePeeExt" to something else? lol.    How about "peerAKIExt"?

File: base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java
Comment (ladycfu) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java]: wrt auths.revocationChecking.enabled=true, I'm still not sure why it would require ldapStore.  I will take a closer look later.

Comment (fmarco76) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java]: Because the method `OCSPEngine.isRevoked()` is invoked when a new connection to OCSP is created. This method uses the LDAPStore to verify the certificate. If LDAPStore is not used, or the CRL verification is disabled I have not verified what happen.

Comment (ladycfu) [reply base/ocsp/src/main/java/com/netscape/cms/ocsp/LDAPStore.java]: ok I just dug slightly deeper and saw that you had overwritten the isRevoked in the OCSPEngine.

**Review by ladycfu — COMMENTED** [review]
Hi Marco, thanks for the update.  I think the only things I have for you now is just to address the couple minor requests I have regarding adding a comment explaining the new param and changing one variable name.

**Review by ladycfu — APPROVED** [review]
Thanks for the update.  LGTM

**Comment by fmarco76** [comment]
@edewata @ladycfu Thanks!

**Pull request merged**
